### ISSUE TITLE
Build FAILURE on PL/Java packaging

### ISSUE THREAD

**Issue by TasMot** [tada/pljava#122]
Running build on Windows 7,  Postgres 9.6, Visual Studio 2017 Community edition

"naraol" is not getting set. How should it be getting set?

Thanks,

Tom 

```
D:\pljava>mvn --version
Apache Maven 3.5.0 (ff8f5e7444045639af65f6095c62210b5713f426; 2017-04-03T15:39:06-04:00)
Maven home: C:\apache-maven-3.5.0\bin\..
Java version: 1.8.0_131, vendor: Oracle Corporation
Java home: C:\Program Files\Java\jdk1.8.0_131\jre
Default locale: en_US, platform encoding: Cp1252
OS name: "windows 7", version: "6.1", arch: "amd64", family: "windows"

D:\pljava>git --version
git version 2.13.0.windows.1

D:\pljava>pg_config
BINDIR = C:/POSTGR~1/pg96/bin
DOCDIR = C:/POSTGR~1/pg96/share/doc/POSTGR~1
HTMLDIR = C:/POSTGR~1/pg96/share/doc/POSTGR~1
INCLUDEDIR = C:/POSTGR~1/pg96/include
PKGINCLUDEDIR = C:/POSTGR~1/pg96/include/POSTGR~1
INCLUDEDIR-SERVER = C:/POSTGR~1/pg96/include/POSTGR~1/server
LIBDIR = C:/POSTGR~1/pg96/lib
PKGLIBDIR = C:/POSTGR~1/pg96/lib/POSTGR~1
LOCALEDIR = C:/POSTGR~1/pg96/share/locale
MANDIR = C:/PostgreSQL/pg96/share/man
SHAREDIR = C:/POSTGR~1/pg96/share/POSTGR~1
SYSCONFDIR = C:/PostgreSQL/pg96/etc/postgresql
PGXS = C:/POSTGR~1/pg96/lib/POSTGR~1/pgxs/src/MAKEFI~1/pgxs.mk
CONFIGURE = '--prefix=C:/msys64/opt/pgbin-build/pgbin/bin/../20170509_0538/build/pg96-9.6.3-1-win64' '--enable-integer-datetimes
' '--enable-thread-safety' '--with-libxml' '--with-libxslt' '--with-ossp-uuid' '--enable-nls' '--with-openssl' '--with-python' '
--with-perl' '--with-tcl' '--with-libraries=/opt/pgbin-build/pgbin/shared/win64/lib:/usr/local/lib:/usr/local/ssl/lib' '--with-i
ncludes=/usr/local/include:/usr/local/ssl/include' 'CFLAGS=-O2 -DMS_WIN64 -I/opt/pgbin-build/pgbin/shared/win64/include'
CC = gcc
CPPFLAGS = -DFRONTEND -I./src/include/port/win32 -DEXEC_BACKEND -I/usr/local/include/libxml2 -I/usr/local/include -I/usr/local/s
sl/include -I../../src/include/port/win32 -DBUILDING_DLL
CFLAGS = -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wfo
rmat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -O2 -DMS_WIN64 -I/opt/pgbin-build/pgbin/shared/win64/incl
ude
CFLAGS_SL =
LDFLAGS = -L../../src/common -Wl,--allow-multiple-definition -Wl,--disable-auto-import -L/usr/local/lib -LC:/msys64/mingw64/lib
-L/opt/pgbin-build/pgbin/shared/win64/lib -L/usr/local/lib -L/usr/local/ssl/lib -Wl,--as-needed
LDFLAGS_EX =
LDFLAGS_SL =
LIBS = -lpgcommon -lpgport -lintl -lxslt -lxml2 -lssl -lcrypto -lz -lws2_32 -lm  -lws2_32
VERSION = PostgreSQL 9.6.3


init:
[loadresource] Do not set property naraol as its length is 0.

package:
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] PostgreSQL PL/Java ................................. SUCCESS [  0.706 s]
[INFO] PL/Java API ........................................ SUCCESS [  1.848 s]
[INFO] PL/Java backend Java code .......................... SUCCESS [  1.202 s]
[INFO] PL/Java backend native code ........................ SUCCESS [  6.600 s]
[INFO] PL/Java Ant tasks .................................. SUCCESS [  0.159 s]
[INFO] PL/Java examples ................................... SUCCESS [  0.453 s]
[INFO] PL/Java packaging .................................. FAILURE [  0.577 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 12.119 s
[INFO] Finished at: 2017-06-12T11:17:12-04:00
[INFO] Final Memory: 66M/1045M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-antrun-plugin:1.7:run (pljava package distribution) on project plj
ava-packaging: An Ant BuildException has occured: The following error occurred while executing this line:
[ERROR] D:\pljava\pljava-packaging\build.xml:259: D:\pljava\pljava-so\target\nar\pljava-so-1.6.0-SNAPSHOT-${naraol}-plugin\lib\$
{naraol}\plugin does not exist.
[ERROR] around Ant part ...<ant dir="D:\pljava\pljava-packaging" target="package"/>... @ 4:59 in D:\pljava\pljava-packaging\targ
et\antrun\build-main.xml
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :pljava-packaging
```

**Comment by jcflack** [comment]
On 06/12/2017 11:41 AM, TasMot wrote:
> Running build on Windows 7,  Postgres 9.6, Visual Studio 2017 Community edition
> 
> "naraol" is not getting set. How should it be getting set?

The pljava-packaging build.xml looks for a file produced
by the pljava-so build (which should already have succeeded
by the time pljava-packaging runs), named history.xml and
found in a long path starting with pljava-so/target/nar
and having somewhere in the middle .../lib/foo/plugin/...
where foo gets picked out as the naraol value. In the first
20 lines or so of pljava-packaging/build.xml.

Can you confirm that you've got a history.xml somewhere
under pljava-so/target/nar and what its full path is, and
perhaps see a reason it didn't satisfy the pattern match?

I might think Windows \ versus / but no, this has all been
tested on Windows before the 1.5.0 release, so it can't be
that easy.

-Chap

**Comment by TasMot** [comment]
Thanks for your help.

In my original post it shows in the Reactor Summary that the pljava-so did run. However; I don't see the history.xml file in the directory. Where do I look in pljava-so pom.xml to see where the history.xml is supposed to be getting created. When I use Eclipse and search the entire directory structure, it doesn't find a reference to that name except to read it in.

Thanks,

Tom

**Comment by TasMot** [comment]
I did this command: mvn clean install -X > build.txt
Attached is the output that was in the build.txt. around line 10690 is the "init:" that starts showing the problem.

[build.txt](https://github.com/tada/pljava/files/1069219/build.txt)

**Comment by jcflack** [comment]
On 06/12/2017 02:43 PM, TasMot wrote:

> I don't see the history.xml file in the directory. Where do I look
> in pljava-so pom.xml to see where the history.xml is supposed
> to be getting created.

You make me nostalgic for the days I thought Maven would be that simple.
:)

The creation of history.xml is entirely a feature of the
nar-maven-plugin, which is used in building pljava-so. You won't
find anything in pom.xml that tells it to do that; it is just
something the plugin does.

Or doesn't do, in your case. That's the odd part. What does
appear under pljava-so/target/nar ?

Perhaps it would be worth re-running mvn with -X and capturing the
output, at least to see what version of nar-maven-plugin it is
pulling in.

-Chap

**Comment by jcflack** [comment]
On 06/12/2017 03:02 PM, TasMot wrote:
> I did this command: mvn clean install -X > build.txt
> Attached is the output that was in the build.txt. around line 10690 is the "init:" that starts showing the problem.
> 
> [build.txt](https://github.com/tada/pljava/files/1069219/build.txt)

Hm, I'll have to get back to this later today.

One thing I notice (probably unrelated) is that you seem to be
building 1.6.0-SNAPSHOT. That's still in very early development
for a (more distant) future release.

I'd recommend trying the current REL1_5_STABLE branch,
a/k/a 1.5.1-SNAPSHOT. That is very soon to become 1.5.1-BETA1.
That is the branch that currently has the updates to work with
PG 9.6 and 10. My plan is to get 1.5.1 out the door and then
merge those changes into the 1.6.0 development ... which is
another way of saying, they're not in that branch yet, so even
if the packaging worked, 1.6.0-SNAPSHOT would probably
give you trouble with pg 9.6.

-Chap

**Comment by TasMot** [comment]
I would gladly do that if you could tell me how. What do I add to the "mvn clean install" to make it build the correct version?

I really appreciate the help.

Tom

**Comment by jcflack** [comment]
On 06/12/2017 04:04 PM, TasMot wrote:
> I would gladly do that if you could tell me how. What do I add
> to the "mvn clean install" to make it build the correct version?

Ah. If you used 'git clone' to get the sources, just use

git checkout REL1_5_STABLE

and then repeat the mvn clean install.

-Chap

**Comment by jcflack** [comment]
Weirdly, from your [build.txt](https://github.com/tada/pljava/files/1069219/build.txt), the `nar-maven-plugin` didn't build the C sources at all, and reported success! Around the same place my build log says

```
[DEBUG]   (f) runtime = dynamic
[DEBUG]   (f) skip = false
[DEBUG]   (f) mavenProject = MavenProject: org.postgresql:pljava-so:1.5.1-SNAPSHOT @ /var/tmp/pljava/pljava-so/pom.xml
[DEBUG]   (f) session = org.apache.maven.execution.MavenSession@6107165
[DEBUG] -- end configuration --
[DEBUG] Added to sourceDirectory: /var/tmp/pljava/pljava-so/src/main
[DEBUG] Added to sourceDirectory: /var/tmp/pljava/pljava-so/src/main
[DEBUG] Added to sourceDirectory: /var/tmp/pljava/pljava-so/src/main
[INFO] Compiling 51 native files
[DEBUG] Creating Library Library: type: plugin
```

Yours says:

```
[DEBUG]   (f) runtime = dynamic
[DEBUG]   (f) skip = false
[DEBUG]   (f) mavenProject = MavenProject: org.postgresql:pljava-so:1.6.0-SNAPSHOT @ D:\pljava\pljava-so\pom.xml
[DEBUG]   (f) session = org.apache.maven.execution.MavenSession@37af1f93
[DEBUG] -- end configuration --
[DEBUG] Added to sourceDirectory: D:\pljava\pljava-so\src\main
[DEBUG] Added to sourceDirectory: D:\pljava\pljava-so\src\main\java
[DEBUG] Added to sourceDirectory: D:\pljava\pljava-so\src\main
[DEBUG] Added to sourceDirectory: D:\pljava\pljava-so\src\main\java
[DEBUG] Added to sourceDirectory: D:\pljava\pljava-so\src\main
[DEBUG] Added to sourceDirectory: D:\pljava\pljava-so\src\main\java
[INFO] Nothing to compile
```

Two things I'm puzzled by at the moment: why java subdirectories got added in your case and not in mine (when the `pljava-so` subproject doesn't have Java sources), and why it then didn't find anything to compile.

Both logs show the same version of `nar-maven-plugin` (as they should, because that is specified in `pom.xml`), but I notice we're not using the same version of Maven itself. (I haven't updated it for a while, still using 3.3.3 here).

Have you had any different outcome using the 1.5.1-SNAPSHOT sources?

**Comment by jcflack** [comment]
One other thing I noticed in passing: it looks like you're using a PostgreSQL build that was done using MinGW-w64, but you're building PL/Java with MSVC? I don't think that's the cause of your current problems, but I suspect it may not load successfully even after getting over the build hurdle.

To work with a MinGW-w64 PostgreSQL build, you'll probably have better results following the PL/Java [MinGW-w64 build procedure](http://tada.github.io/pljava/build/mingw64.html).

**Comment by TasMot** [comment]
[build.txt](https://github.com/tada/pljava/files/1071436/build.txt)

I think it went backwards on the build. now the pljava-so step is failing. 

I installed the PostgreSQL build from BigSQL. That is why it is using the MinGW-w64 build. I don't have the tools to build Postgres (or time). 

I read another post that said that the MinGW build is does not work/is not supported on Windows. 

I'm actually just looking for a version of PLJava that works with Postgres 9.6. I am trying to hook up a business rules engine (Drools) to Postgres. Since the business rules get built into a Java class, the PLJava extension seemed perfect.

Thanks for your help.

**Comment by jcflack** [comment]
On 06/13/2017 09:01 AM, TasMot wrote:

> I read another post that said that the MinGW build is
> does not work/is not supported on Windows. 

Hmm, if you can find that post again can you link me to it?
It sounds like they didn't look at the PL/Java 1.5.0 release notes:

> Reported build issues for specific platforms have been resolved, with
> new platform-specific build documentation for OS X, Solaris, Ubuntu,
> Windows MSVC, and Windows MinGW-w64.

> I'm actually just looking for a version of PLJava that works with
> Postgres 9.6.

I know BigSQL offers a package of PL/Java 1.5.0 that you can install
with 'pgc install', so I assume once 1.5.1 is released (working with
PostgreSQL 9.6 and 10), they will have that available too. I'm sorry
you've had trouble building from source.

-Chap

**Comment by TasMot** [comment]
I still had it open: http://www.postgresql-archive.org/Build-Fails-on-Windows-7-td5888020.html

**Comment by jcflack** [comment]
On 06/13/2017 10:10 AM, TasMot wrote:
> I still had it open: http://www.postgresql-archive.org/Build-Fails-on-Windows-7-td5888020.html 

Ah. That was the thread from a year ago February that ended with
having both MSVC and MinGW-w64 building supported in time for
the 1.5.0 release.  :)

-Chap

**Comment by TasMot** [comment]
OK, so what I had to do: I uninstalled Postgres 9.6, deleted everything to completely get rid of it, installed Postgres 9.5 and then implemented pljava 1.5. So far, this seems to be working. I am trying to get the "Hello, World" example coded and working.

Do you have any idea what the ETA is for the PLJava that matches Postgres 9.6?

Thanks,

Tom

**Comment by jcflack** [comment]
On 06/14/2017 09:19 AM, TasMot wrote:
> OK, so what I had to do: I uninstalled Postgres 9.6, deleted everything to completely get rid of it, installed Postgres 9.5 and then implemented pljava 1.5. So far, this seems to be working. I am trying to get the "Hello, World" example coded and working.
> 
> Do you have any idea what the ETA is for the PLJava that matches Postgres 9.6?

I want to announce 1.5.1-BETA1 this week. Downstream packagers
might make 1.5.1-BETA1 packages available, depending on their
policies. How long it takes between beta and final 1.5.1 depends
on what issues people discover, but the changes from 1.5.0 are
not numerous so I expect the beta to be fairly painless. It seems
reasonable to allow about a month of beta, to allow any problems
in the field to be reported.

Did you Maven-build 1.5.0, or use the BigSQL prebuilt version?

-Chap

**Comment by TasMot** [comment]
I used the BigSQL prebuilt version.

I would be willing to retry the build though. I think I will need to deal with getting the correct configuration set-up.  I installed/uninstalled/reinstalled a lot of things trying to get it to work. I am fairly new to Maven. I used mostly ANT in the past.

Thanks again for your help.

Tom

**Comment by jcflack** [comment]
> Weirdly, from your [build.txt](https://github.com/tada/pljava/files/1069219/build.txt), the `nar-maven-plugin` didn't build the C sources at all, and reported success!
> ...
> Both logs show the same version of `nar-maven-plugin` (as they should, because that is specified in `pom.xml`), but I notice we're not using the same version of Maven itself. (I haven't updated it for a while, still using 3.3.3 here).

I thought that might indicate some Maven 3.5 weirdness, but I have now built with that version here with no issues, so that strange nar plugin behavior may remain a mystery unless it can be caught in the act.

**Comment by jcflack** [comment]
PL/Java 1.5.1 is released, supporting PostgreSQL 11 back to 8.2. Closing.
